worker/peergrouper/worker.go

```diff
@@ -675,9 +680,21 @@ func (w *pgWorker) updateReplicaSet() (map[string]*replicaset.Member, error) {
 	if err := setHasVote(added, true); err != nil {
 		return nil, errors.Annotate(err, "adding new voters")
 	}
+
+	// Figure out if we are running on the mongo primary.
+	controllerId := w.config.ControllerIdGetter()
+	isPrimary, err := info.isPrimary(controllerId)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, errors.Annotatef(err, "determining primary status of controller %q", controllerId)
+	}
+	logger.Debugf("controller node %q primary: %v", controllerId, isPrimary)
 	// Currently k8s controllers do not support HA, so only update
 	// the replicaset config if HA is enabled and there is a change.
-	if w.config.SupportsHA && desired.isChanged {
+	// Only controllers corresponding with the mongo primary should
+	// update the replicaset, otherwise there will be a race since
+	// a diff needs to be calculated so the changes can be applied
+	// one at a time.
+	if isPrimary && w.config.SupportsHA && desired.isChanged {
```

Comment:
Below this block is:
```go
if err := setHasVote(removed, false); err != nil {
		return nil, errors.Annotate(err, "removing non-voters")
}
```
For consistency I'd also do only this on the primary.

Reply:
At the bottom of this method, I would also raise the logging level for controllers still alive, from which voting rights have been removed. That seems like something to surface.